src/main/java/org/apache/maven/plugin/compiler/TestCompilerMojo.java

```diff
@@ -478,58 +472,173 @@ protected void patchModulesForTestDependencies(
+                  &&
+                  // Is it different from main module name
+                  !mainModuleName.equals( moduleName )
+            ) 
             {
-                // Get module name
-                JavaModuleDescriptor moduleDesc = result.getPathElements().get( path );
-                String moduleName = ( moduleDesc != null ) ? moduleDesc.name() : null;
+            
+                // Add --add-reads <moduleName>=ALL-UNNAMED
+                compilerArgs.add( "--add-reads" );
+                StringBuilder sbAddReads = new StringBuilder();
+                sbAddReads.append( moduleName ).append( "=ALL-UNNAMED" );
+                compilerArgs.add( sbAddReads.toString() );
                 
-                if ( 
-                      // Is it a modularized compile elements?
-                      ( path != null ) && ( moduleName != null )
-                      &&
-                      // Is it different from main module name
-                      !mainModuleName.equals( moduleName )
-                ) 
+                // Add compile classpath if needed
+                if ( !listModuleTestPaths.isEmpty() )
                 {
-                
-                    // Add compile classpath
                     // Yes, add it as patch module
                     List<String> listPath = patchModules.get( moduleName );
                     // Make sure it is initialized
                     if ( listPath == null ) 
                     {
                         listPath = new ArrayList<>();
                         patchModules.put( moduleName, listPath );
-                    }
+                    } 
                     
                     // Add test compile path but not main module target
-                    listPath.addAll( testPath );
-                    // Remove main target path
-                    listPath.remove( mainModuleTarget );
-                    
+                    listPath.addAll( listModuleTestPaths );
                 }
+                
             }
             
-            if ( pathElements == null ) 
+        }
+
+        if ( pathElements == null ) 
+        {
+            pathElements = new LinkedHashMap<>( result.getPathElements().size() );
+            for ( Entry<File, JavaModuleDescriptor> entry : result.getPathElements().entrySet() ) 
             {
-                pathElements = new LinkedHashMap<>( result.getPathElements().size() );
-                for ( Entry<File, JavaModuleDescriptor> entry : result.getPathElements().entrySet() ) 
-                {
-                    pathElements.put( entry.getKey().getAbsolutePath(), entry.getValue() );
-                }
+                pathElements.put( entry.getKey().getAbsolutePath(), entry.getValue() );
             }
         }
     }
 
+    /**
+     * Get module test path element from module path element
+     * @param modulePathElt
+     * @return
+     */
+    private File getModuleTestPathElt( File modulePathElt )
```

Comment:
This methods feels a bit like best guessing. It is better to loop over all the reactorProjects to find the Paths that belong to the same Maven Project (and because of that to the same module)